### ISSUE TITLE
in_tail: Fluentd v1.12.3 fails to handle file rotations on Windows

### ISSUE THREAD

**Issue by fujimotos** [fluent/fluentd#3365]
**Describe the bug**

After a few file rotations, `in_tail` stops tracking log files.

**To Reproduce**

1.  Run Fluentd with the attached configuration.

2. Rotate the file twice using cmd.exe

   ```cmd
   % cd /opt/td-agent/
   % echo aiueo >> test.log
   % del test.log
   % echo aiueo >> test.log
   % del test.log
   ```

3. Now Fluentd just gets stuck, keeping `test.log` in `DeletePending` status forever.

**Expected behavior**

File rotation should work properly.

**Your Environment**

- Windows Server 2016
- td-agent v4.1.1
- Fluentd v1.12.3

**Your Configuration**

```xml
<source>
@type tail
 path /opt/td-agent/test.log
 tag test.log
 <parse>
   @type none
 </parse>
</source>

<match test**>
  @type stdout
</match>
```

**Additional context**

* This issue is evidenlty occurring since #3196.

**Comment by ashie** [comment]
@cosmo0920 Could you take look this?
Feel free to reassign me if you can't.

**Comment by cosmo0920** [comment]
Sure. I'll tackle on this in the next week. :eyes:

**Comment by cosmo0920** [comment]
I'd confirmed that the current master (b6ba1c27db9a1f1b736b20c1c5018ed6cd88a01e) can be stucked with the attached configuration.
And I found that taking over 5 seconds to adding lines into rotated file can be reproduce this issue:
```cmd
% del test.log
del Adding a line after long time (over 5 seconds)
% echo aiueo >> test.log
```

**Comment by cosmo0920** [comment]
I'd confirmed that this issue is also happened on Linux. Not for Windows specific.

**Comment by AmeerAssi** [comment]
thanks @cosmo0920.
@fujimotos mentioned that the root cause of this issue: https://github.com/fluent/fluentd/issues/3340 might be the issue here.

we are using Fluentd v.1.11.2, do you know if its reproducing with this version?

**Comment by cosmo0920** [comment]
> we are using Fluentd v.1.11.2, do you know if its reproducing with this version?

This version is not reproducible this issue. This issue is only for Fluentd v1.12.x.

**Comment by cosmo0920** [comment]
I've sent a PR to fix this issue: https://github.com/fluent/fluentd/pull/3391

And I have never reproduced this issue with https://github.com/fluent/fluentd/pull/3391.

**Comment by ashie** [comment]
Note for the mechanism of this bug:
Since hash key of TargetInfo is based on whole of it's contents, it doesn't match with existing `TailWatcher` in `@tails` on `refresh_watchers` if inode is changed.

https://github.com/fluent/fluentd/blob/5773739bdb817a96d88cae2f49318031b7c03293/lib/fluent/plugin/in_tail.rb#L363-L374

Thus TailWatcher for the path stays seated in `@tails` forever, so that `refresh_watcher` can't create a new TailWatcher for the path.
On the other hand, since `refresh_watcher` is called on every 60 seconds, rotation will be handled by `on_rotate` (not by `refresh_watcher`) when a file is create before next call of `refresh_watcher`, so following a new file is worked correctly.

**Comment by ashie** [comment]
Fixed via #3395
